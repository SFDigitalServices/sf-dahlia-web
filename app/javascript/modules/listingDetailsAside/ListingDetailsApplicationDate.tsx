import React from "react"
import { RailsListing } from "../listings/SharedHelpers"
import { ApplicationStatus, ApplicationStatusType, Icon, t } from "@bloom-housing/ui-components"
import { localizedFormat } from "../../util/languageUtil"
import dayjs from "dayjs"
import { LISTING_TYPE_FIRST_COME_FIRST_SERVED } from "../constants"
import { useFeatureFlag } from "../../hooks/useFeatureFlag"
import { isFcfsSalesListing, isOpen } from "../../util/listingUtil"
import bloomTheme from "../../../../tailwind.config"
import { Message } from "@bloom-housing/ui-seeds"

export interface ListingDetailsApplicationDateProps {
  listing: RailsListing
}

export const formatTime = (time: string) => {
  const formattedTime = dayjs(time).format("h:mm")
  const hour = Number(dayjs(time).format("H"))
  // \u00A0 is a non-breaking space
  const suffix = hour >= 12 ? "\u00A0PM" : "\u00A0AM"
  return `${formattedTime}${suffix}`
}

const StatusMessage = ({ isClosed, message }: { isClosed: boolean; message: string }) => {
  return (
    <Message
      fullwidth
      className="justify-start leading-5"
      variant={isClosed ? "alert" : "primary"}
      customIcon={<Icon fill={isClosed ? "red-700" : ""} symbol="clock" size="medium" />}
    >
      {message}
    </Message>
  )
}

const getStatusLottery = (listing: RailsListing) => {
  const isClosed = !isOpen(listing)
  const datetime = listing.Application_Due_Date
  return (
    <StatusMessage
      isClosed={isClosed}
      message={t(
        isClosed
          ? "listingDetails.applicationsClosed.withDateTime"
          : "listingDetails.applicationsDeadline.withDateTime",
        {
          date: localizedFormat(datetime, "LL"),
          time: formatTime(datetime),
        }
      )}
    />
  )
}

const getStatusFcfs = (listing: RailsListing) => {
  const isClosed = !listing.Accepting_Online_Applications
  const datetime = listing.Application_Start_Date_Time
  const showDateTime = new Date(listing.Application_Start_Date_Time) > new Date()

  const getMessage = () => {
    if (isClosed) {
      return t("listingDetails.applicationsClosed")
    } else {
      return showDateTime
        ? t("listingDetails.applicationsOpen.withDateTime", {
            date: localizedFormat(datetime, "LL"),
            time: formatTime(datetime),
          })
        : t("listingDetails.applicationsOpen")
    }
  }

  return <StatusMessage isClosed={isClosed} message={getMessage()} />
}

const ListingDetailsStatus = ({ listing }: { listing: RailsListing }) => {
  return listing.Listing_Type === LISTING_TYPE_FIRST_COME_FIRST_SERVED
    ? getStatusFcfs(listing)
    : getStatusLottery(listing)
}

export const ListingDetailsApplicationDate = ({ listing }: ListingDetailsApplicationDateProps) => {
  const isApplicationOpen = listing && isOpen(listing)

  const { unleashFlag: isSalesFcfsEnabled } = useFeatureFlag("FCFS", false)

  return (
    <div className="w-full mb-8 md:mb-0">
      {isSalesFcfsEnabled && isFcfsSalesListing(listing) ? (
        <ListingDetailsStatus listing={listing} />
      ) : (
        <div className="w-full mb-8 md:mb-0">
          <ApplicationStatus
            className="place-content-center"
            content={t(
              isApplicationOpen
                ? "listingDetails.applicationDeadline.open"
                : "listingDetails.applicationDeadline.closed",
              {
                date: localizedFormat(listing.Application_Due_Date, "ll"),
                time: dayjs(listing.Application_Due_Date).format("h:mm A"),
              }
            )}
            iconColor={!isApplicationOpen && bloomTheme.theme.colors.red["700"]}
            status={isApplicationOpen ? ApplicationStatusType.Open : ApplicationStatusType.Closed}
          />
        </div>
      )}
    </div>
  )
}
