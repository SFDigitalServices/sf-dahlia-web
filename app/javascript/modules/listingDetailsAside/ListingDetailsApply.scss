.fcfs-bmr-how-to-apply {
  padding: var(--seeds-spacer-content) !important;
}

.fcfs-bmr-how-to-apply__list {
  padding-top: var(--seeds-spacer-content) !important;
  padding-bottom: var(--seeds-spacer-content) !important;
}

.fcfs-bmr-how-to-apply {
  h4.text__underline-weighted {
    margin-bottom: 0px;
  }
}
