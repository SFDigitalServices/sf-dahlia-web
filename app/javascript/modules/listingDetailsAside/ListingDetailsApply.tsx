import React, { useState } from "react"
import {
  AppearanceStyleType,
  Button,
  Heading,
  LinkButton,
  MultiLineAddress,
  OrDivider,
  SidebarBlock,
  t,
} from "@bloom-housing/ui-components"
import { RailsListing } from "../listings/SharedHelpers"
import {
  acceptingPaperApplications,
  isOpen,
  isRental,
  isSale,
  paperApplicationURLs,
} from "../../util/listingUtil"
import { localizedFormat, renderInlineWithInnerHTML } from "../../util/languageUtil"

export interface ListingDetailsApplyProps {
  listing: RailsListing
}

export const ListingDetailsApply = ({ listing }: ListingDetailsApplyProps) => {
  const [paperApplicationsOpen, setPaperApplicationsOpen] = useState(false)

  if (!isOpen(listing)) return null

  const isListingRental = isRental(listing)

  const acceptingPaperApps = acceptingPaperApplications(listing)

  const ordinalHeader = (ordinal: number, title: string) => {
    return (
      <Heading priority={4} className={"text-gray-900 text-lg -mt-2 mb-3"}>
        <span className={"text-blue-600 mr-2"}>{ordinal}</span>
        {title}
      </Heading>
    )
  }

  const howToApplyBlock = (
    <SidebarBlock title={t("listings.apply.howToApply")}>
      {!isListingRental && (
        <>
          <p className={"mb-4"}>
            {renderInlineWithInnerHTML(
              t("listings.apply.fulfillEligibilityRequirements", {
                url: "https://sfmohcd.org/homebuyer-program-eligibility",
              })
            )}
          </p>
          <p className={"mb-4"}>{t("listings.apply.eligibilityRequirementDescription")}</p>
        </>
      )}
      <Button styleType={AppearanceStyleType.primary} className={"w-full"} transition={true}>
        {t("label.applyOnline")}
      </Button>
      {acceptingPaperApps && (
        <>
          <OrDivider bgColor={"white"} />
          {ordinalHeader(1, t("listings.apply.getAPaperApplication"))}
          {t("listings.apply.paperApplicationsMustBeMailed")}
          <Button
            className={"w-full mt-4"}
            transition={true}
            icon={"arrowDown"}
            iconPlacement={"right"}
            onClick={() => setPaperApplicationsOpen(!paperApplicationsOpen)}
          >
            {t("label.downloadApplication", {
              lang: "",
            })}
          </Button>
          {paperApplicationsOpen && acceptingPaperApplications && (
            <div className={"flex w-full items-center justify-center flex-col"}>
              {paperApplicationURLs(isListingRental).map((app) => {
                return (
                  <LinkButton href={app.fileURL} unstyled className={"m-0 pt-4 no-underline"}>
                    {app.languageString}
                  </LinkButton>
                )
              })}
            </div>
          )}
        </>
      )}
    </SidebarBlock>
  )

  const submitPaperApplicationBlocks = (
    <>
      <SidebarBlock className={"bg-blue-200"}>
        {ordinalHeader(2, t("listings.apply.submitAPaperApplication"))}
        {t("listings.apply.includeAnEnvelope")}
      </SidebarBlock>
      <SidebarBlock
        className={"bg-blue-200"}
        style={"sidebarSubHeader"}
        title={t("listings.apply.sendByUsMail")}
      >
        <div className={"mb-2 text-gray-900 text-base"}>
          <MultiLineAddress
            address={{
              placeName: listing.Application_Organization,
              street: listing.Application_Street_Address,
              city: listing.Application_City,
              zipCode: listing.Application_Postal_Code,
              state: listing.Application_State,
            }}
          />
        </div>
        {t("listings.apply.applicationsMustBeReceivedByDeadline")}
      </SidebarBlock>
      <SidebarBlock className={"bg-blue-200"}>
        {`${t("listings.apply.doNotApplyOnlineAndPaper")} ${t(
          "listings.apply.doNotDoMultipleApplications"
        )}`}
      </SidebarBlock>
    </>
  )

  const needHelpBlock = (
    <SidebarBlock title={t("listings.apply.needHelp")}>
      {isListingRental && (
        <div className={"mb-4"}>{t("listings.apply.visitAHousingCounselor")}</div>
      )}
      <LinkButton
        transition={true}
        newTab={true}
        href={"https://www.homeownershipsf.org/application-assistance-for-homebuyers/"}
        className={"w-full"}
      >
        {t("housingCounselor.findAHousingCounselor")}
      </LinkButton>
    </SidebarBlock>
  )

  const expectedMoveInDateBlock = (
    <SidebarBlock title={t("listings.expectedMoveinDate")}>
      {localizedFormat(listing.Expected_Move_in_Date, "MMMM YYYY")}
    </SidebarBlock>
  )

  return (
    <>
      {howToApplyBlock}
      {acceptingPaperApps && submitPaperApplicationBlocks}
      {needHelpBlock}
      {isSale(listing) && listing.Expected_Move_in_Date && expectedMoveInDateBlock}
    </>
  )
}
