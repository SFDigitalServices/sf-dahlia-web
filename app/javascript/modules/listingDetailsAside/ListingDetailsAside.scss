.fcfs-aside {
    --card-border-radius: rounded-none;
    --card-background-color: var(--bloom-color-gray-200);
    --card-footer-background-color: var(--bloom-color-gray-200);
    --card-border-width: border-b-1;

    .aside-card-header {
        font-family: var(--seeds-font-alt-sans);
        font-size: var(--seeds-font-size-lg);
        font-weight: var(-seeds-font-weight-semibold);
        line-height: var(--seeds-line-height-6);
    }

    .aside-card-section {
        font-family: var(--seeds-font-sans);
        font-size: var(--seeds-font-size-sm);
        font-weight: var(--seeds-font-weight-regular);
        line-height: var(--seeds-line-height-5);
    }
}
