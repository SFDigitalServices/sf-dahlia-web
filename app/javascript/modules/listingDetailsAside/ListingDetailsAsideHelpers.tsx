import { t } from "@bloom-housing/ui-components"
import React from "react"

export const SeeThisUnitMLS = ({ url }: { url: string }) => {
  return (
    <p className="mt-1">
      <a href={url} target="_blank">
        {t("listings.process.seeThisUnitOnMls")}
      </a>
    </p>
  )
}
