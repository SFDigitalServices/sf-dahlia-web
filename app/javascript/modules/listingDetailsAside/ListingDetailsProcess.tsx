import React from "react"
import { getEventNote, RailsListing } from "../listings/SharedHelpers"
import dayjs from "dayjs"
import { isFcfsSalesListing, isRental, isSale } from "../../util/listingUtil"
import {
  EventSection,
  Contact,
  t,
  ExpandableSection,
  SidebarBlock,
  Desktop,
} from "@bloom-housing/ui-components"
import { getTranslatedString, localizedFormat, renderInlineMarkup } from "../../util/languageUtil"
import { ListingDetailsLotteryPreferenceLists } from "./ListingDetailsLotteryPreferenceLists"
export interface ListingDetailsProcessProps {
  listing: RailsListing
  isApplicationOpen: boolean
}

const WhatToExpect = () => {
  return (
    <div>
      <ExpandableSection
        content={t("emailer.submissionConfirmation.applicantsWillBeContacted")}
        expandableContent={
          <>
            <p>{t("f2ReviewTerms.p3")}</p>
            <p className={"mt-2 mb-2"}>{t("label.whatToExpectApplicationChosen")}</p>
          </>
        }
        strings={{
          title: t("label.whatToExpect"),
          readMore: t("label.showMore"),
          readLess: t("label.showLess"),
          buttonAriaLabel: t("listings.whatToExpect.showMore.aria"),
        }}
        priority={2}
      />
    </div>
  )
}

export const ListingDetailsProcess = ({
  listing,
  isApplicationOpen,
}: ListingDetailsProcessProps) => {
  const isListingSale = isSale(listing)
  const isListingRental = isRental(listing)

  return (
    <>
      {!!listing.Lottery_Date &&
        dayjs(listing.Lottery_Date) > dayjs() &&
        !listing.LotteryResultsURL &&
        isApplicationOpen && (
          <div className="border-b border-gray-400 md:border-b-0">
            <EventSection
              events={[
                {
                  dateString: localizedFormat(listing.Lottery_Date, "LL"),
                  timeString: dayjs(listing.Lottery_Date).tz().format("hh:mma"),
                  note: getEventNote({
                    City: listing.Lottery_City,
                    Street_Address: listing.Lottery_Street_Address,
                    Venue: getTranslatedString(
                      listing.Lottery_Venue,
                      "Lottery_Venue__c",
                      listing.translations
                    ),
                  }),
                },
              ]}
              headerText={t("listings.process.publicLottery")}
              sectionHeader={true}
            />
          </div>
        )}
      {!isFcfsSalesListing(listing) && <WhatToExpect />}
      <ListingDetailsLotteryPreferenceLists
        listing={listing}
        isApplicationOpen={isApplicationOpen}
      />
      {
        // For sales listings, Contact info is in the See the Unit component
        (listing.Leasing_Agent_Email ||
          listing.Leasing_Agent_Name ||
          listing.Leasing_Agent_Phone ||
          listing.Office_Hours ||
          listing.Leasing_Agent_Title) &&
          isListingRental && (
            <div className="border-b border-gray-400 md:border-b-0 last:border-b-0">
              <Contact
                sectionTitle={t("contactAgent.contact")}
                priority={2}
                contactAddress={{
                  street: listing.Leasing_Agent_Street,
                  city: listing.Leasing_Agent_City,
                  state: listing.Leasing_Agent_State,
                  zipCode: listing.Leasing_Agent_Zip,
                }}
                additionalInformation={
                  listing.Office_Hours
                    ? [
                        {
                          title: t("contactAgent.officeHours"),
                          content: (
                            <span className="translate">
                              {renderInlineMarkup(
                                getTranslatedString(
                                  listing.Office_Hours,
                                  "Office_Hours__c",
                                  listing.translations
                                )
                              )}
                            </span>
                          ),
                        },
                      ]
                    : undefined
                }
                contactEmail={listing.Leasing_Agent_Email}
                contactName={listing.Leasing_Agent_Name}
                contactPhoneNumber={
                  listing.Leasing_Agent_Phone
                    ? t("listings.call", { phoneNumber: listing.Leasing_Agent_Phone })
                    : undefined
                }
                contactPhoneNumberNote={t("contactAgent.dueToHighCallVolume")}
                contactTitle={getTranslatedString(
                  listing.Leasing_Agent_Title,
                  "Leasing_Agent_Title__c",
                  listing.translations
                )}
                contactTitleClassname="translate"
                strings={{
                  email: t("label.emailAddress"),
                  getDirections: t("label.getDirections"),
                }}
              />
            </div>
          )
      }
      <Desktop>
        {isListingSale && (
          <div className="border-b border-gray-400 md:border-b-0 last:border-b-0">
            <SidebarBlock title={t("listings.housingProgram")} priority={2}>
              <a href={`https://sfmohcd.org/for-buyers`} target="_blank" className="text-base">
                {t("listings.belowMarketRate")}
              </a>
            </SidebarBlock>
          </div>
        )}
        {isApplicationOpen && (
          <div className="border-b border-gray-400 md:border-b-0 last:border-b-0">
            <SidebarBlock priority={2}>
              <p>{`${t("t.listingUpdated")}: ${localizedFormat(
                listing.LastModifiedDate,
                "LL"
              )}`}</p>
            </SidebarBlock>
          </div>
        )}
      </Desktop>
    </>
  )
  /* TODO: Bloom prop changes <DownloadLotteryResults resultsDate={"January 1st, 2022"} pdfURL={""} /> */
}
