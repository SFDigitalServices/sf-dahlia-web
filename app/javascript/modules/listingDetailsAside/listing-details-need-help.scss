.listing-details-need-help {
  padding: var(--seeds-spacer-content) !important;
  border-color: var(--bloom-color-gray-400) !important;
}
