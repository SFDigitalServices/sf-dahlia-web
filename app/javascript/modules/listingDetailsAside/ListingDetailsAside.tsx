import React from "react"
import { ListingDetailItem, t } from "@bloom-housing/ui-components"
import { RailsListing } from "../listings/SharedHelpers"
import { ListingDetailsInfoSession } from "./ListingDetailsInfoSession"
import { ListingDetailsProcess } from "./ListingDetailsProcess"
import { isOpen } from "../../util/listingUtil"
import { ListingDetailsApply } from "./ListingDetailsApply"
import { ListingDetailsApplicationDate } from "./ListingDetailsApplicationDate"
import { ListingDetailsLotteryResults } from "../listingDetailsLottery/ListingDetailsLotteryResults"

export interface ListingDetailsSidebarProps {
  listing: RailsListing
  imageSrc: string
}

export const ListingDetailsAside = ({ listing, imageSrc }: ListingDetailsSidebarProps) => {
  const isApplicationOpen = isOpen(listing)

  return (
    <ListingDetailItem
      imageAlt={""}
      imageSrc={imageSrc}
      title={t("listings.process.header")}
      subtitle={t("listings.process.subheader")}
      hideHeader={true}
      desktopClass="header-hidden"
    >
      <aside className="w-full static md:absolute md:right-0 md:w-1/3 md:top-0 sm:w-2/3 md:ml-2 h-full md:border border-solid bg-white">
        <div className="hidden md:block">
          <ListingDetailsApplicationDate isApplicationOpen={isApplicationOpen} listing={listing} />
          <ListingDetailsLotteryResults listing={listing} />
          {isApplicationOpen && <ListingDetailsInfoSession listing={listing} />}
          <ListingDetailsApply listing={listing} />
          <ListingDetailsProcess listing={listing} />
        </div>
      </aside>
    </ListingDetailItem>
  )
}
