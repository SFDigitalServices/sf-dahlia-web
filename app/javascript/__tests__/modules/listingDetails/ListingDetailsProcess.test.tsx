import React from "react"
import { render } from "@testing-library/react"
import { ListingDetailsProcess } from "../../../modules/listingDetailsAside/ListingDetailsProcess"
import { lotteryCompleteRentalListing } from "../../data/RailsRentalListing/listing-rental-lottery-complete"

describe("ListingDetailsProcess", () => {
  it("renders leasing agent section if exists", () => {
    window.matchMedia = jest.fn().mockImplementation((query) => {
      return {
        matches: true,
        media: query,
        onchange: null,
        addListener: jest.fn(),
        removeListener: jest.fn(),
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      }
    })
    
    const { asFragment } = render(
      <ListingDetailsProcess listing={lotteryCompleteRentalListing} isApplicationOpen={true} />
    )

    expect(asFragment()).toMatchSnapshot()
  })
})
