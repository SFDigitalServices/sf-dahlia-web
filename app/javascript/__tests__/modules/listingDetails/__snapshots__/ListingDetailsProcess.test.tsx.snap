// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListingDetailsProcess renders leasing agent section if exists 1`] = `
Array [
  <section
    className="aside-block"
  >
    <h4
      className="text-caps-underline"
    >
      What to Expect
    </h4>
    <div
      className="text-tiny text-gray-750"
    >
      <span>
        Applicants will be contacted by the agent in lottery rank order until vacancies are filled.
      </span>
      <div
        className="mt-2"
      >
        <div>
          <button
            aria-expanded={false}
            className="button is-unstyled m-0 no-underline has-toggle"
            onClick={[Function]}
            type="button"
          >
            Read More
          </button>
        </div>
      </div>
    </div>
  </section>,
  <section
    className="aside-block"
  >
    <h4
      className="text-caps-underline undefined"
    >
      Contact
    </h4>
    <p
      className="text-xl"
    >
      Kathryn Kelley
    </p>
    <p
      className="text-gray-700"
    >
      Leasing Agent
    </p>
    <p
      className="mt-3"
    >
      <a
        href="tel:Call 4158728106"
      >
        <span
          className="ui-icon ui-medium pr-2"
          data-test-id={null}
        >
          <svg
            fill="#0077DA"
            viewBox="0 0 32 32"
          >
            <path
              d="M22 20c-2 2-2 4-4 4s-4-2-6-4-4-4-4-6 2-2 4-4-4-8-6-8-6 6-6 6c0 4 4.109 12.109 8 16s12 8 16 8c0 0 6-4 6-6s-6-8-8-6z"
            />
          </svg>
        </span>
        Call 415-872-8106
      </a>
    </p>
    <p
      className="text-sm text-gray-700"
    >
      Due to high call volume you may hear a message.
    </p>
    <p
      className="my-3"
    >
      <a
        href="mailto:ldavis3@eqr.com"
      >
        <span
          className="ui-icon ui-medium pr-2"
          data-test-id={null}
        >
          <svg
            fill="#0077DA"
            viewBox="0 0 32 32"
          >
            <path
              d="M29.091 27.637h-26.182c-1.606 0-2.909-1.304-2.909-2.909v-17.454c0-1.606 1.304-2.909 2.909-2.909h26.182c1.606 0 2.909 1.304 2.909 2.909v17.454c0 1.606-1.304 2.909-2.909 2.909zM27.637 24.728l-8.338-7.296-3.299 2.933-3.299-2.933-8.338 7.296h23.272zM2.909 23.272l8.182-7.272-8.182-7.272v14.546zM4.363 7.272l11.637 10.182 11.637-10.182h-23.272zM29.091 8.728l-8.182 7.272 8.182 7.272v-14.546z"
            />
          </svg>
        </span>
        Email
      </a>
    </p>
    <h3
      className="text-caps-tiny undefined"
    >
      Office Hours
    </h3>
    <div
      className="text-gray-800 text-tiny markdown"
    >
      Monday - Friday 10:00 AM - 6:00 PM
    </div>
  </section>,
]
`;
